File: 프로그래머스/lv2/42839. 소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#include <string>
#include <map>
#include <iostream>
using namespace std;

int visited[7];
map<unsigned long long, int> M;
bool isPrime(string temp){
    //cout << temp << "\n";
    
    if(temp == "" || stoi(temp) < 2)
        return false;
    int tn = stoi(temp);
    for(int i = 2; i < tn; i++){
        if(tn % i == 0)
            return false;
    }
    return true;
}

void go(string temp, string numbers){
    if(isPrime(temp)){
        M[stoi(temp)]++;    
    }
    
    for(int i = 0; i < numbers.length(); i++){
        if(!visited[i]){
            visited[i] = 1;
            go(temp + numbers[i], numbers);
            visited[i] = 0;
        }
    }
}

int solution(string numbers) {
    int answer = 0;
    string temp = "";
    go(temp, numbers);
    
    // for(int i = 0; i <= 9999999; i++){
    //     M[i] = 0;
    // }
    
    for(int i = 0; i <= 9999999; i++){
        if(M[i] >= 1){
            cout << i << "\n";
            answer++;
        }
    }
    return answer;
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    static boolean[] visited;
    static int N;
    static int ret = 0;
    static Set<Integer> existN = new HashSet<>();
    public int solution(String numbers) {
        int answer = 0;
        N = numbers.length();
        visited = new boolean[numbers.length()];
        String temp = "";
        go(0, numbers, temp);
        answer = ret;
        return answer;
    }
    
    public boolean isPrime(String temp){
        int num = Integer.parseInt(temp);
        if(existN.contains(num)){
            return false;    
        }
        existN.add(num);

        if(num < 2)
            return false;
        for(int i = 2; i < num; i++){
            if(num % i == 0){
                return false;
            }
        }
        return true;
    }
    
    public void go(int depth, String numbers, String temp){
        if(temp.length() > N)
            return;
        if(depth != 0){
            if(isPrime(temp)){
                ret++;
            }
        }
        
        for(int i = 0; i < N; i++){
            if(!visited[i]){
                visited[i] = true;
                go(depth+1, numbers, temp + numbers.charAt(i));
                visited[i] = false;
            }
        }
    }
}